Created script to provision and destroy development IPA cluster (#722)
greenplum-db/pxf: dev/IPA.md

# Creating a multi-node Hadoop cluster with Kerberos provided by an IPA server

If you need to create a multinode Hadoop cluster secured by IPA in Google Cloud Platform you can run the
script [`ipa-cluster.bash` bash script](ipa-cluster.bash) that does the following:
* uses `terraform` templates from 
[concourse/terraform/ipa-multinode-hadoop/templates](../concourse/terraform/ipa-multinode-hadoop/templates) 
directory along with [local.tfvars](../concourse/terraform/ipa-multinode-hadoop/templates/local.tfvars) file to
spin up GCP instances for an IPA server, hadoop namenodes and datanodes.
* uses the [`ipa-multinode-hadoop` Ansible play](../concourse/ansible/ipa-multinode-hadoop) 
to configure and start the Hadoop cluster.
* generates TLS private keys, self-signed certificates and pkcs12 keystores for the Hadoop nodes.

> * creates PXF configuration server under `$PXF_BASE` and deployes `*-site.xml` files and keytabs there

Reviewer: ```suggestion
* creates PXF configuration server under `$PXF_BASE` and deploys `*-site.xml` files and keytabs there
```

```

## Creating the cluster
Execute the following command to create a new cluster:
```
ipa-cluster.bash --create
```
The script will:
 * validate that the required software is present and the environment variables are set
 * generate necessary additional `terraform` artifacts in `concourse/terraform/ipa-multinode-hadoop/templates` directory
 * spin up GCP instances for the Hadoop cluster
 * apply ansible configuration steps to configure the Hadoop cluster
 * add a private key to access GCP VMs as `~/.ssh/[username]`
 * add SSH configurations to `~/.ssh/config` file
 * create a PXF configuration server `$PXF_BASE/servers/hdfs-ipa`
 * print out messages instructing the user to:
   * enable constrained delegation property in their `$PXF_BASE/servers/hdfs-ipa/pxf-site.xml`
   * add an extra property to `$PXF_BASE/servers/hdfs-ipa/hdfs-site.xml`
   * update their `/etc/hosts` file
   * create or update their `/etc/krb5.conf` file

One manual step for now is to go and add the following property to the `$PXF_BASE/servers/hdfs-ipa/hdfs-site.conf`:
```
    <property>
        <name>dfs.client.use.datanode.hostname</name>
        <value>true</value>
    </property>
```
   
## Outcome

> If the script executes successfully, and you follow the post-execution steps, you will have:

Reviewer: ```suggestion 
If the script executes successfully and you follow the post-execution steps, you will have:
```

## Working with the cluster
To list HDFS directories of the new cluster accessible by PXF service user, run:
```
kinit -kt ${PXF_BASE}/servers/hdfs-ipa/pxf.service.keytab porter@[YOUR REALM]
hdfs --config ${PXF_BASE}/servers/hdfs-ipa dfs -ls /
```
To list HDFS directories of the new cluster accessible by Haddop admin user, run:
```
kinit -kt ${PXF_BASE}/servers/hdfs-ipa/hadoop.user.keytab stout@[YOUR REALM]
hdfs --config ${PXF_BASE}/servers/hdfs-ipa dfs -ls /
```

## Cleanup

IMPORTANT: the cluster provisioned by the script is not yet integrated with the CPP reaper. 
This means you will need to manually issue a command that destroys all GCP artifacts created by the script.

Run the following command to destroy the cluster:
```
ipa-cluster.bash --destroy
```
This command invokes `terraform destroy` to cleanup all provisioned resources. 
Inspect the output and double check in GCP console that all the VM instances named like `ccp-[username]-*` are destroyed.

You might also want to manually delete your PXF configuration server `$PXF_BASE/servers/hdfs-ipa` 
and any modifications you have done to:
- `/etc/hosts` file
- `/etc/krb5.conf` file
- `~/.ssh/config` file

> - remove `~/.ssh/[username]` and `~/.ssh/[username].pub` files which has a private key to now destroyed GCP VM instances 

Reviewer: ```suggestion
- remove `~/.ssh/[username]` and `~/.ssh/[username].pub` files which has a private key to the GCP VM instances that were just destroyed
```
